### ISSUE TITLE
Enhancement: [no-confusing-non-null-assertion] should check !instanceof

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8795]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-confusing-non-null-assertion/

### Description

I propose that the rule should flag on `x !instanceof y`, (and, also, IMO on `x! instanceof y`, but this part is more debatable).

### Fail

```TypeScript
declare const x: any;
// logs false
console.log(x !instanceof SyntaxError);
// also logs false, same AST
console.log(x! instanceof SyntaxError);
```

### Pass

```TypeScript
declare const x: any;
// logs true
console.log(!(x instanceof SyntaxError));
```

### Additional Info

I have encountered this twice independently in the wild at work.
Perhaps part of the confusion comes from the fact that in some other languages, boolean operators can behave more fluently, for example, in Python:
```python
zero = 0
zero is not False # True, equivalent to not (zero is False)
zero is (not False) # False
```
so it's intuitive to think that one could write 
```typescript
x !instanceof y
```
to mean 
```typescript
!(x instanceof y)
```
and not realize that TS is doing you dirty by permitting that syntax at all (which is not valid JS).

Note that prettier autocorrects to the `x! instanceof` form instead of `x !instanceof` form. But either way it's yikes.

**Comment by kirkwaiblinger** [comment]
Note that this is achievable `no-restricted-syntax` using `"BinaryExpression[operator=instanceof] > TSNonNullExpression"` as long as you don't mind the fact that you'll flag on `x! instanceof y` as well, not just `x !instanceof y`.

**Comment by Josh-Cena** [comment]
...and of course `!in` too
